# <u>**Part 9**



In [11]:
import numpy as np

### **Exercise 81**

Using Numpy generate all dates for mondays from January and February 2020 as an array and print result to the console.

In [9]:
np.arange('2020-01-06', '2020-03', dtype='datetime64[D]')[::7]

array(['2020-01-06', '2020-01-13', '2020-01-20', '2020-01-27',
       '2020-02-03', '2020-02-10', '2020-02-17', '2020-02-24'],
      dtype='datetime64[D]')

### **Exercise 82**

The followings arrays are given

A = np.array(['001', '002', '003'], dtype=np.str) <br>
B = np.array(['XC', 'YC', 'ZC'], dtype=np.str)


Merge them together 

In [12]:
A = np.array(['001', '002', '003'], dtype=str)
B = np.array(['XC', 'YC', 'ZC'], dtype=str)

In [13]:
np.char.add(A, B)

array(['001XC', '002YC', '003ZC'], dtype='<U5')

### **Exercise 83**

The following array is given

A = np.array(['1', '2', '3'], dtype=np.str)

Add '000' (3 zeros) at the beginning of each element of this array


In [15]:
A = np.array(['1', '2', '3'], dtype=str)

In [16]:
print(np.char.add('000', A))
print(np.char.rjust(A, 4, fillchar='0'))
print(np.char.zfill(A, 4))

['0001' '0002' '0003']
['0001' '0002' '0003']
['0001' '0002' '0003']


### **Exercise 84**

The following array is given

[['PLW CDR 11B TEN', 'AMC LPP'], <br>
['CDR PKO KGH', 'CCC QMK']]

Split each element of this array by space character 

In [17]:
A = np.array([['PLW CDR 11B TEN', 'AMC LPP'],
              ['CDR PKO KGH', 'CCC QMK']], dtype=str)

In [18]:
np.char.split(A)

array([[list(['PLW', 'CDR', '11B', 'TEN']), list(['AMC', 'LPP'])],
       [list(['CDR', 'PKO', 'KGH']), list(['CCC', 'QMK'])]], dtype=object)

### **Exercise 85**

The following array is given

[['#summer#time#mood'], <br>
['#sport#time']]

Replace all # characters with spaces and then strip all unnecessary white characters around the text.


In [35]:
A = np.array([['#summer#time#mood'],
              ['#sport#time']])
A

array([['#summer#time#mood'],
       ['#sport#time']], dtype='<U17')

In [36]:
print(np.char.strip(np.char.replace(A, '#', ' ')))

[['summer time mood']
 ['sport time']]


In [37]:
A = np.char.replace(A, '#', ' ')
A

array([[' summer time mood'],
       [' sport time']], dtype='<U17')

In [38]:
A = np.char.strip(A, ' ')
A

array([['summer time mood'],
       ['sport time']], dtype='<U17')

### **Exercise 86**

The following array is given

[['#summer#time#mood', '#vibe'], <br>
['#sport#time', '#good#time'], <br>
['#event#summer', '#fast#move']] <br>

Count the number of occurrences of the word 'time' in each element of this array.



In [39]:
A = np.array([['#summer#time#mood', '#vibe'],
              ['#sport#time', '#good#time'],
              ['#event#summer', '#fast#move']])

In [41]:
np.char.count(A, 'time')

array([[1, 0],
       [1, 1],
       [0, 0]])

### **Exercise 87**

The following text is given

text = """ALIOR    PLALIOR00045    88 860 000    1 386 216 000    0,891    2,16    14 <br>
CCC    PLCCC0000016    27 918 000    1 292 603 400    0,831    5,28    42 <br>
CDPROJEKT    PLOPTTC00011    67 348 000    22 864 646 000    14,702    7,39    7 <br>
CYFRPLSAT    PLCFRPT00013    275 301 000    6 854 994 900    4,408    1,17    14 <br>
DINOPL    PLDINPL00011    47 937 000    8 916 282 000    5,733    9,13    12 <br>
JSW    PLJSW0000015    52 636 000    716 902 320    0,461    1,51    24 <br>
KGHM    PLKGHM000017    136 410 000    9 881 540 400    6,354    4,78    8 <br>
LOTOS    PLLOTOS00025    86 543 000    5 609 717 260    3,607    2,91    16 <br>
LPP    PLLPP0000011    1 306 000    7 444 200 000    4,787    1,43    19 <br>
MBANK    PLBRE0000012    12 997 000    2 830 746 600    1,820    0,42    24 <br>
ORANGEPL    PLTLKPL00017    647 357 000    4 285 503 340    2,756    1,16    13 <br>
PEKAO    PLPEKAO00016    176 379 000    9 619 710 660    6,185    5,27    9 <br>
PGE    PLPGER000010    796 776 000    3 561 588 720    2,290    2,88    18 <br>
PGNIG    PLPGNIG00014    1 624 608 000    6 072 784 704    3,905    1,56    12 <br>
PKNORLEN    PLPKN0000018    289 049 000    17 701 360 760    11,382    12,44    8 <br>
PKOBP    PLPKO0000016    857 593 000    18 807 014 490    12,093    10,49    9 <br>
PLAY    LU1642887738    114 151 000    3 696 209 380    2,377    1,47    16 <br>
PZU    PLPZU0000011    568 305 000    17 515 160 100    11,262    6,64    6 <br>
SANPL    PLBZ00000044    33 207 000    5 213 499 000    3,352    1,91    18 <br>
TAURONPE    PLTAURN00011    1 043 590 000    1 252 308 000    0,805    1,21    33""" 

The columns mean: Instrument | ISIN code | Package | Package (PLN) | Share in portfolio (%) | Share in trading shares and PDAs per session (%) | Average spread per session

Split text into lines. Then split each line by the tab character \t and create an array (without headers) containing the data

In [53]:
text = """ALIOR	PLALIOR00045	88 860 000	1 386 216 000	0,891	2,16	14
CCC	PLCCC0000016	27 918 000	1 292 603 400	0,831	5,28	42
CDPROJEKT	PLOPTTC00011	67 348 000	22 864 646 000	14,702	7,39	7
CYFRPLSAT	PLCFRPT00013	275 301 000	6 854 994 900	4,408	1,17	14
DINOPL	PLDINPL00011	47 937 000	8 916 282 000	5,733	9,13	12
JSW	PLJSW0000015	52 636 000	716 902 320	0,461	1,51	24
KGHM	PLKGHM000017	136 410 000	9 881 540 400	6,354	4,78	8
LOTOS	PLLOTOS00025	86 543 000	5 609 717 260	3,607	2,91	16
LPP	PLLPP0000011	1 306 000	7 444 200 000	4,787	1,43	19
MBANK	PLBRE0000012	12 997 000	2 830 746 600	1,820	0,42	24
ORANGEPL	PLTLKPL00017	647 357 000	4 285 503 340	2,756	1,16	13
PEKAO	PLPEKAO00016	176 379 000	9 619 710 660	6,185	5,27	9
PGE	PLPGER000010	796 776 000	3 561 588 720	2,290	2,88	18
PGNIG	PLPGNIG00014	1 624 608 000	6 072 784 704	3,905	1,56	12
PKNORLEN	PLPKN0000018	289 049 000	17 701 360 760	11,382	12,44	8
PKOBP	PLPKO0000016	857 593 000	18 807 014 490	12,093	10,49	9
PLAY	LU1642887738	114 151 000	3 696 209 380	2,377	1,47	16
PZU	PLPZU0000011	568 305 000	17 515 160 100	11,262	6,64	6
SANPL	PLBZ00000044	33 207 000	5 213 499 000	3,352	1,91	18
TAURONPE	PLTAURN00011	1 043 590 000	1 252 308 000	0,805	1,21	33"""

In [54]:
lines = text.splitlines()
lines = [line.split('\t') for line in lines]
result = np.array(lines, dtype=str)
print(result)

[['ALIOR' 'PLALIOR00045' '88 860 000' '1 386 216 000' '0,891' '2,16' '14']
 ['CCC' 'PLCCC0000016' '27 918 000' '1 292 603 400' '0,831' '5,28' '42']
 ['CDPROJEKT' 'PLOPTTC00011' '67 348 000' '22 864 646 000' '14,702'
  '7,39' '7']
 ['CYFRPLSAT' 'PLCFRPT00013' '275 301 000' '6 854 994 900' '4,408' '1,17'
  '14']
 ['DINOPL' 'PLDINPL00011' '47 937 000' '8 916 282 000' '5,733' '9,13'
  '12']
 ['JSW' 'PLJSW0000015' '52 636 000' '716 902 320' '0,461' '1,51' '24']
 ['KGHM' 'PLKGHM000017' '136 410 000' '9 881 540 400' '6,354' '4,78' '8']
 ['LOTOS' 'PLLOTOS00025' '86 543 000' '5 609 717 260' '3,607' '2,91' '16']
 ['LPP' 'PLLPP0000011' '1 306 000' '7 444 200 000' '4,787' '1,43' '19']
 ['MBANK' 'PLBRE0000012' '12 997 000' '2 830 746 600' '1,820' '0,42' '24']
 ['ORANGEPL' 'PLTLKPL00017' '647 357 000' '4 285 503 340' '2,756' '1,16'
  '13']
 ['PEKAO' 'PLPEKAO00016' '176 379 000' '9 619 710 660' '6,185' '5,27' '9']
 ['PGE' 'PLPGER000010' '796 776 000' '3 561 588 720' '2,290' '2,88' '18']
 ['PGNIG' 'PL

### **Exercise 88**

The following array is given

result = np.array([['ALIOR', 'PLALIOR00045', '88 860 000', '1 386 216 000', '0,891', '2,16', '14'], <br>
       ['CCC', 'PLCCC0000016', '27 918 000', '1 292 603 400', '0,831', '5,28', '42'], <br>
       ['CDPROJEKT', 'PLOPTTC00011', '67 348 000', '22 864 646 000', '14,702', '7,39', '7'], <br>
       ['CYFRPLSAT', 'PLCFRPT00013', '275 301 000', '6 854 994 900', '4,408', '1,17', '14'], <br>
       ['DINOPL', 'PLDINPL00011', '47 937 000', '8 916 282 000', '5,733', '9,13', '12'], <br>
       ['JSW', 'PLJSW0000015', '52 636 000', '716 902 320', '0,461', '1,51', '24'], <br>
       ['KGHM', 'PLKGHM000017', '136 410 000', '9 881 540 400', '6,354', '4,78', '8'], <br>
       ['LOTOS', 'PLLOTOS00025', '86 543 000', '5 609 717 260', '3,607', '2,91', '16'], <br>
       ['LPP', 'PLLPP0000011', '1 306 000', '7 444 200 000', '4,787', '1,43', '19'], <br>
       ['MBANK', 'PLBRE0000012', '12 997 000', '2 830 746 600', '1,820', '0,42', '24'], <br>
       ['ORANGEPL', 'PLTLKPL00017', '647 357 000', '4 285 503 340', '2,756', '1,16', '13'], <br>
       ['PEKAO', 'PLPEKAO00016', '176 379 000', '9 619 710 660', '6,185', '5,27', '9'], <br>
       ['PGE', 'PLPGER000010', '796 776 000', '3 561 588 720', '2,290', '2,88', '18'], <br>
       ['PGNIG', 'PLPGNIG00014', '1 624 608 000', '6 072 784 704', '3,905', '1,56', '12'], <br>
       ['PKNORLEN', 'PLPKN0000018', '289 049 000', '17 701 360 760', '11,382', '12,44', '8'], <br>
       ['PKOBP', 'PLPKO0000016', '857 593 000', '18 807 014 490', '12,093', '10,49', '9'], <br>
       ['PLAY', 'LU1642887738', '114 151 000', '3 696 209 380', '2,377', '1,47', '16'], <br>
       ['PZU', 'PLPZU0000011', '568 305 000', '17 515 160 100', '11,262', '6,64', '6'], <br>
       ['SANPL', 'PLBZ00000044', '33 207 000', '5 213 499 000', '3,352', '1,91', '18'], <br>
       ['TAURONPE', 'PLTAURN00011', '1 043 590 000', '1 252 308 000', '0,805', '1,21', '33']], dtype='<U14')

Remove all spaces in the text and replace all commas with periods. 

In [50]:
result = np.array([['ALIOR', 'PLALIOR00045', '88 860 000', '1 386 216 000', '0,891', '2,16', '14'],
       ['CCC', 'PLCCC0000016', '27 918 000', '1 292 603 400', '0,831', '5,28', '42'],
       ['CDPROJEKT', 'PLOPTTC00011', '67 348 000', '22 864 646 000', '14,702', '7,39', '7'],
       ['CYFRPLSAT', 'PLCFRPT00013', '275 301 000', '6 854 994 900', '4,408', '1,17', '14'],
       ['DINOPL', 'PLDINPL00011', '47 937 000', '8 916 282 000', '5,733', '9,13', '12'],
       ['JSW', 'PLJSW0000015', '52 636 000', '716 902 320', '0,461', '1,51', '24'],
       ['KGHM', 'PLKGHM000017', '136 410 000', '9 881 540 400', '6,354', '4,78', '8'],
       ['LOTOS', 'PLLOTOS00025', '86 543 000', '5 609 717 260', '3,607', '2,91', '16'],
       ['LPP', 'PLLPP0000011', '1 306 000', '7 444 200 000', '4,787', '1,43', '19'],
       ['MBANK', 'PLBRE0000012', '12 997 000', '2 830 746 600', '1,820', '0,42', '24'],
       ['ORANGEPL', 'PLTLKPL00017', '647 357 000', '4 285 503 340', '2,756', '1,16', '13'],
       ['PEKAO', 'PLPEKAO00016', '176 379 000', '9 619 710 660', '6,185', '5,27', '9'],
       ['PGE', 'PLPGER000010', '796 776 000', '3 561 588 720', '2,290', '2,88', '18'],
       ['PGNIG', 'PLPGNIG00014', '1 624 608 000', '6 072 784 704', '3,905', '1,56', '12'],
       ['PKNORLEN', 'PLPKN0000018', '289 049 000', '17 701 360 760', '11,382', '12,44', '8'],
       ['PKOBP', 'PLPKO0000016', '857 593 000', '18 807 014 490', '12,093', '10,49', '9'],
       ['PLAY', 'LU1642887738', '114 151 000', '3 696 209 380', '2,377', '1,47', '16'],
       ['PZU', 'PLPZU0000011', '568 305 000', '17 515 160 100', '11,262', '6,64', '6'],
       ['SANPL', 'PLBZ00000044', '33 207 000', '5 213 499 000', '3,352', '1,91', '18'],
       ['TAURONPE', 'PLTAURN00011', '1 043 590 000', '1 252 308 000', '0,805', '1,21', '33']], dtype='<U14')

In [52]:
result = np.char.replace(result, ' ', '')
result = np.char.replace(result, ',', '.')
print(result)

[['ALIOR' 'PLALIOR00045' '88860000' '1386216000' '0.891' '2.16' '14']
 ['CCC' 'PLCCC0000016' '27918000' '1292603400' '0.831' '5.28' '42']
 ['CDPROJEKT' 'PLOPTTC00011' '67348000' '22864646000' '14.702' '7.39' '7']
 ['CYFRPLSAT' 'PLCFRPT00013' '275301000' '6854994900' '4.408' '1.17' '14']
 ['DINOPL' 'PLDINPL00011' '47937000' '8916282000' '5.733' '9.13' '12']
 ['JSW' 'PLJSW0000015' '52636000' '716902320' '0.461' '1.51' '24']
 ['KGHM' 'PLKGHM000017' '136410000' '9881540400' '6.354' '4.78' '8']
 ['LOTOS' 'PLLOTOS00025' '86543000' '5609717260' '3.607' '2.91' '16']
 ['LPP' 'PLLPP0000011' '1306000' '7444200000' '4.787' '1.43' '19']
 ['MBANK' 'PLBRE0000012' '12997000' '2830746600' '1.820' '0.42' '24']
 ['ORANGEPL' 'PLTLKPL00017' '647357000' '4285503340' '2.756' '1.16' '13']
 ['PEKAO' 'PLPEKAO00016' '176379000' '9619710660' '6.185' '5.27' '9']
 ['PGE' 'PLPGER000010' '796776000' '3561588720' '2.290' '2.88' '18']
 ['PGNIG' 'PLPGNIG00014' '1624608000' '6072784704' '3.905' '1.56' '12']
 ['PKNORLEN' 

### **Exercise 89**

result = np.array([['ALIOR', 'PLALIOR00045', '88860000', '1386216000', '0.891', '2.16', '14'], <br>
['CCC', 'PLCCC0000016', '27918000', '1292603400', '0.831', '5.28', '42'], <br>
['CDPROJEKT', 'PLOPTTC00011', '67348000', '22864646000', '14.702', '7.39', '7'], <br>
['CYFRPLSAT', 'PLCFRPT00013', '275301000', '6854994900', '4.408', '1.17', '14'], <br>
['DINOPL', 'PLDINPL00011', '47937000', '8916282000', '5.733', '9.13', '12'], <br>
['JSW', 'PLJSW0000015', '52636000', '716902320', '0.461', '1.51', '24'], <br>
['KGHM', 'PLKGHM000017', '136410000', '9881540400', '6.354', '4.78', '8'], <br>
['LOTOS', 'PLLOTOS00025', '86543000', '5609717260', '3.607', '2.91', '16'], <br>
['LPP', 'PLLPP0000011', '1306000', '7444200000', '4.787', '1.43', '19'], <br>
['MBANK', 'PLBRE0000012', '12997000', '2830746600', '1.820', '0.42', '24'], <br>
['ORANGEPL', 'PLTLKPL00017', '647357000', '4285503340', '2.756', '1.16', '13'], <br>
['PEKAO', 'PLPEKAO00016', '176379000', '9619710660', '6.185', '5.27', '9'], <br>
['PGE', 'PLPGER000010', '796776000', '3561588720', '2.290', '2.88', '18'], <br>
['PGNIG', 'PLPGNIG00014', '1624608000', '6072784704', '3.905', '1.56', '12'], <br>
['PKNORLEN', 'PLPKN0000018', '289049000', '17701360760', '11.382', '12.44', '8'], <br>
['PKOBP', 'PLPKO0000016', '857593000', '18807014490', '12.093', '10.49', '9'], <br>
['PLAY', 'LU1642887738', '114151000', '3696209380', '2.377', '1.47', '16'], <br>
['PZU', 'PLPZU0000011', '568305000', '17515160100', '11.262', '6.64', '6'], <br>
['SANPL', 'PLBZ00000044', '33207000', '5213499000', '3.352', '1.91', '18'], <br>
['TAURONPE', 'PLTAURN00011', '1043590000', '1252308000', '0.805', '1.21', '33']], dtype='<U12')

Extract all rows with the company name starting with the letter 'P' and assign to the *stocks_startswith_P* variable.

In [55]:
result = np.array([['ALIOR', 'PLALIOR00045', '88 860 000', '1 386 216 000', '0,891', '2,16', '14'],
       ['CCC', 'PLCCC0000016', '27 918 000', '1 292 603 400', '0,831', '5,28', '42'],
       ['CDPROJEKT', 'PLOPTTC00011', '67 348 000', '22 864 646 000', '14,702', '7,39', '7'],
       ['CYFRPLSAT', 'PLCFRPT00013', '275 301 000', '6 854 994 900', '4,408', '1,17', '14'],
       ['DINOPL', 'PLDINPL00011', '47 937 000', '8 916 282 000', '5,733', '9,13', '12'],
       ['JSW', 'PLJSW0000015', '52 636 000', '716 902 320', '0,461', '1,51', '24'],
       ['KGHM', 'PLKGHM000017', '136 410 000', '9 881 540 400', '6,354', '4,78', '8'],
       ['LOTOS', 'PLLOTOS00025', '86 543 000', '5 609 717 260', '3,607', '2,91', '16'],
       ['LPP', 'PLLPP0000011', '1 306 000', '7 444 200 000', '4,787', '1,43', '19'],
       ['MBANK', 'PLBRE0000012', '12 997 000', '2 830 746 600', '1,820', '0,42', '24'],
       ['ORANGEPL', 'PLTLKPL00017', '647 357 000', '4 285 503 340', '2,756', '1,16', '13'],
       ['PEKAO', 'PLPEKAO00016', '176 379 000', '9 619 710 660', '6,185', '5,27', '9'],
       ['PGE', 'PLPGER000010', '796 776 000', '3 561 588 720', '2,290', '2,88', '18'],
       ['PGNIG', 'PLPGNIG00014', '1 624 608 000', '6 072 784 704', '3,905', '1,56', '12'],
       ['PKNORLEN', 'PLPKN0000018', '289 049 000', '17 701 360 760', '11,382', '12,44', '8'],
       ['PKOBP', 'PLPKO0000016', '857 593 000', '18 807 014 490', '12,093', '10,49', '9'],
       ['PLAY', 'LU1642887738', '114 151 000', '3 696 209 380', '2,377', '1,47', '16'],
       ['PZU', 'PLPZU0000011', '568 305 000', '17 515 160 100', '11,262', '6,64', '6'],
       ['SANPL', 'PLBZ00000044', '33 207 000', '5 213 499 000', '3,352', '1,91', '18'],
       ['TAURONPE', 'PLTAURN00011', '1 043 590 000', '1 252 308 000', '0,805', '1,21', '33']], dtype='<U14')

In [58]:
stocks_startswith_P = result[np.char.startswith(result[:, 0], 'P')]
print(stocks_startswith_P)

[['PEKAO' 'PLPEKAO00016' '176 379 000' '9 619 710 660' '6,185' '5,27' '9']
 ['PGE' 'PLPGER000010' '796 776 000' '3 561 588 720' '2,290' '2,88' '18']
 ['PGNIG' 'PLPGNIG00014' '1 624 608 000' '6 072 784 704' '3,905' '1,56'
  '12']
 ['PKNORLEN' 'PLPKN0000018' '289 049 000' '17 701 360 760' '11,382'
  '12,44' '8']
 ['PKOBP' 'PLPKO0000016' '857 593 000' '18 807 014 490' '12,093' '10,49'
  '9']
 ['PLAY' 'LU1642887738' '114 151 000' '3 696 209 380' '2,377' '1,47' '16']
 ['PZU' 'PLPZU0000011' '568 305 000' '17 515 160 100' '11,262' '6,64' '6']]


### **Exercise 90**

The following array is given

stocks_startswith_P = np.array([['PEKAO', 'PLPEKAO00016', '176379000', '9619710660', '6.185', '5.27', '9'], <br>
['PGE', 'PLPGER000010', '796776000', '3561588720', '2.290', '2.88', '18'], <br>
['PGNIG', 'PLPGNIG00014', '1624608000', '6072784704', '3.905', '1.56', '12'], <br>
['PKNORLEN', 'PLPKN0000018', '289049000', '17701360760', '11.382', '12.44', '8'], <br>
['PKOBP', 'PLPKO0000016', '857593000', '18807014490', '12.093', '10.49', '9'], <br>
['PLAY', 'LU1642887738', '114151000', '3696209380', '2.377', '1.47', '16'], <br>
['PZU', 'PLPZU0000011', '568305000', '17515160100', '11.262', '6.64', '6']],dtype='<U12')

Calculate the total share of companies whose name starts with the letter 'P'. Round the result to two decimal places 

In [59]:
stocks_startswith_P = np.array([['PEKAO', 'PLPEKAO00016', '176379000', '9619710660', '6.185', '5.27', '9'],
       ['PGE', 'PLPGER000010', '796776000', '3561588720', '2.290', '2.88', '18'],
       ['PGNIG', 'PLPGNIG00014', '1624608000', '6072784704', '3.905', '1.56', '12'],
       ['PKNORLEN', 'PLPKN0000018', '289049000', '17701360760', '11.382', '12.44', '8'],
       ['PKOBP', 'PLPKO0000016', '857593000', '18807014490', '12.093', '10.49', '9'],
       ['PLAY', 'LU1642887738', '114151000', '3696209380', '2.377', '1.47', '16'],
       ['PZU', 'PLPZU0000011', '568305000', '17515160100', '11.262', '6.64', '6']], dtype='<U12')

In [62]:
np.round(stocks_startswith_P[:, 4].astype(float).sum(), 2)

49.49